components/service/glean/src/main/java/mozilla/components/service/glean/storages/TimingDistributionsStorageEngine.kt

```diff
@@ -0,0 +1,243 @@
+ * us that this could leak, the application context lives as long as the application and this
+ * object. For this reason, we should be safe to suppress the IDE warning.
+ */
+@SuppressLint("StaticFieldLeak")
+internal object TimingDistributionsStorageEngine : TimingDistributionsStorageEngineImplementation()
+
+internal open class TimingDistributionsStorageEngineImplementation(
+    override val logger: Logger = Logger("glean/TimingDistributionsStorageEngine")
+) : GenericScalarStorageEngine<TimingDistributionData>() {
+
+    override fun deserializeSingleMetric(metricName: String, value: Any?): TimingDistributionData? {
+        return try {
+            (value as? String)?.let {
+                TimingDistributionData.fromJsonString(it)
+            }
+        } catch (e: org.json.JSONException) {
+            null
+        }
+    }
+
+    override fun serializeSingleMetric(
+        userPreferences: SharedPreferences.Editor?,
+        storeName: String,
+        value: TimingDistributionData,
+        extraSerializationData: Any?
+    ) {
+        val json = value.toJsonObject()
+        userPreferences?.putString(storeName, json.toString())
+    }
+
+    /**
+     * Accumulate value for the provided metric.
+     *
+     * @param metricData the metric information for the timing distribution
+     * @param sample the value to accumulate
+     */
+    @Synchronized
+    fun accumulate(
+        metricData: CommonMetricData,
+        sample: Long,
+        timeUnit: TimeUnit
+    ) {
+        val newTD = TimingDistributionData(category = metricData.category, name = metricData.name,
+            timeUnit = timeUnit)
+        newTD.accumulate(sample)
+        super.recordScalar(metricData, newTD, null) { currentValue, newValue ->
+            currentValue?.let {
+                it.accumulate(newValue.sum)
+                it
+            } ?: newValue
+        }
+    }
+
+    /**
+     * Get a snapshot of the stored data as a JSON object.
+     *
+     * @param storeName the name of the desired store
+     * @param clearStore whether or not to clearStore the requested store
+     *
+     * @return the [JSONObject] containing the recorded data.
+     */
+    override fun getSnapshotAsJSON(storeName: String, clearStore: Boolean): Any? {
+        return getSnapshot(storeName, clearStore)?.let { dataMap ->
+            val jsonObj = JSONObject()
+            dataMap.forEach {
+                jsonObj.put(it.key, it.value.toJsonObject())
+            }
+            return jsonObj
+        }
+    }
+}
+
+@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+data class TimingDistributionData(
+    val category: String,
+    val name: String,
+    val bucketCount: Int = DEFAULT_BUCKET_COUNT,
+    val range: List<Long> = listOf(DEFAULT_RANGE_MIN, DEFAULT_RANGE_MAX),
+    val histogramType: HistogramType = HistogramType.Exponential,
+    val values: MutableMap<String, Long> = mutableMapOf(),
+    var underflow: Long = 0,
```

Comment:
What are `underflow` and `overflow` used for?

Reply:
They were defined in the schema so I added them here, I'll remove them since they aren't currently being used and are listed as 'optional' in the schema definition.

Reply:
I think they are supposed to be buckets containing out of range values.

For timings specifically (as opposed to a general histogram), they are always positive, so one could argue all negative values should go in `underflow`.  Or we just provide an API that makes that impossible, or we treat those as errors through the error reporting mechanism.

If we use the proposed "inifinite buckets" where we don't really have an upper bound (which works pretty nicely for exponential histograms, probably less so for linear), than `overflow` would be impossible.

tl;dr: I'm fine with removing these, remembering that we'll probably want them back when we implement general-purpose histograms...